CiviCase - Fix SelectWhereClause for Activity queries

colemanw opened the pull request:
Overview
----------------------------------------
Fixes bugs in SearchKit with selecting case activities, and a new regression in Advanced Search (picked up by unit test) caused by  918e583a146876fb42932a1d959796fd710b7b74.

Before
----------------------------------------
- Crash if selecting case activities with no permissions for CiviCase.
- "Error - Array to string conversion" in `CRM_Contact_BAO_Query`.

After
----------------------------------------
SelectWhereClause applied properly. Search runs without error (so says the unit test).

colemanw committed:
APIv4 - Don't crash if permission denied to join

colemanw committed:
CiviCase - Add selectWhereClause for activities

demeritcowboy commented on CRM/Case/BAO/Case.php:
```diff
@@ -3019,33 +3019,66 @@ public static function buildOptions($fieldName, $context = NULL, $props = []) {
+
+    // No CiviCase access
+    if (!$viewAllCases && !$viewMyCases) {
+      $clauses['id'][] = 'IS NULL';
```
I think maybe the problem is the query ends up looking like: `((a.id IN (SELECT activity_id FROM civicrm_activity_contact WHERE contact_id IN (SELECT id FROM civicrm_contact WHERE is_deleted != 1))) AND (a.id NOT IN (SELECT activity_id FROM civicrm_case_activity) OR a.id IN (SELECT activity_id FROM civicrm_case_activity WHERE case_id IS NULL AND case_id IN (SELECT id FROM civicrm_case WHERE is_deleted = 0))))`

in particular the part `a.id IN (SELECT activity_id FROM civicrm_case_activity WHERE case_id IS NULL`, because that's never true.

demeritcowboy replied:
Although the other part of the `OR` seems to kick in so maybe that's not the issue. Nevermind.

colemanw committed:
DAO - Handle OR clauses in getDynamicFkAclClauses, improve sql formatting for readability

Follow-up to 918e583a146876fb42932a1d959796fd710b7b74 - this correctly joins sub-arrays using OR

colemanw committed:
Fix tests not setting permissions properly

colemanw committed:
Fix Activity ACL clause generation

Regression was caused by  918e583a146876fb42932a1d959796fd710b7b74 updating DAO::getSelectWhereClause
but not CRM_Contact_BAO_Query::generatePermissionClause. Turns out the code was being duplicated and
the latter could just call the former.

demeritcowboy commented:
I've configured the online demo here so that if you log in as the demo user and go to civicrm/user, it crashes trying to run sql WHERE id Array.

The crash isn't pre-existing, but it's possible there was something else wrong before. Looking.

colemanw commented:
@demeritcowboy and that includes my latest commit [b404968](https://github.com/civicrm/civicrm-core/pull/27722/commits/b4049685d1645666d9674c192e1b14dfbea4f339) (that was supposed to fix that problem)

colemanw commented:
@demeritcowboy tests are passing! 🎉 
What about the issue you spotted? Still there?

demeritcowboy commented:
Good job tracking stuff down. I must have been using the previous demo.

I've put merge-ready since it seems like there's a lot going on here and I don't do a lot of stuff with civi acls.

The test thing is weird if there's a case id that floats around across tests causing regular activities to become case-ified, which is how I'm interpreting the code comments. But I suppose that doesn't need to be fixed here.

colemanw commented:
> The test thing is weird if there's a case id that floats around across tests causing regular activities to become case-ified, which is how I'm interpreting the code comments

Your interpretation of my comment is correct. Whether my comment is correct is another question... but that's my best guess as to what's going on. Anyway, not today's problem.

demeritcowboy commented:
It was bugging me. For sendEmailBasic, the changes to the test seem correct given that that actually is a case-y test because it uses case tokens in the email, so needs civicase. Ditto the campaign one. But then for the other testSendEmailBasicWithoutAnyTokens the actual problem is that teardown doesn't clear civicrm_case_activity, so when the acl check checks that table there IS an activity id that matches, it just happens to also be the same activity id from the previous test, even though it's a different (non-case) activity.

If I add `'civicrm_case_activity', 'civicrm_case_contact', 'civicrm_case',` to teardown, and remove the changes in testSendEmailBasicWithoutAnyTokens, it's all good. I think that's better because then it's a better teardown, and then the non-case test is better since it's testing with minimal permissions.

demeritcowboy commented:
For the campaign one there's two choices: Change getCaseEmailTaskForm to getContactEmailTaskForm and remove the other changes, or just leave it with the changes. It's not really a case test but since it calls the case task it needs civicase.

colemanw committed:
ActivityTest - Fix tearDown to prevent test leakage

colemanw commented:
@demeritcowboy your idea for the 1s test worked! I added the teardown and removed the workaround.
And your idea for the 2nd test also worked! I did nothing, and the test still passes :)

MegaphoneJon commented:
This breaks the hook contract when calling `selectWhereClause()` on activities.  @seamuslee001 @JoeMurray I'm working on a fix for activity type ACLs, but I'm not sure that the solution is to change the extensions.

agileware-justin commented on ext/civi_case/civi_case.php:
```diff
@@ -17,3 +17,22 @@ function civi_case_civicrm_managed(&$entities, $modules) {
+ *
+ * @implements CRM_Utils_Hook::selectWhereClause
+ */
+function civi_case_civicrm_selectWhereClause($entityName, &$clauses, $userId, $conditions) {
```
Just reported a regression for this change and specifically this function, see https://lab.civicrm.org/dev/core/-/issues/4769
